blender: Switch to OpenColorIO 2.x

rastertail opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Adds derivations for pystring and opencolorio2. Right now, OCIO2 is compiled without GPU rendering support, but it should be possible to expose it as an optional feature if we want to.

###### Motivation for this change

Fixes color management in Blender 2.93 / #126726. Should probably be backported to 21.05.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

rastertail committed:
maintainers: Add rytone

rastertail committed:
pystring: init at 1.1.3

Required for OpenColorIO 2.x

rastertail committed:
opencolorio2: init at 2.0.1

rastertail committed:
blender: Switch to opencolorio2

OPNA2608 commented on pkgs/development/libraries/opencolorio/default.nix:
```diff
@@ -1,47 +1,28 @@
     description = "A color management framework for visual effects and animation";
     license = licenses.bsd3;
-    maintainers = [ maintainers.goibhniu ];
+    maintainers = [ maintainers.rytone ];
```
@cillianderoiste, do you want to maintain the 2.x version of this derivation as well?

OPNA2608 commented on pkgs/development/libraries/pystring/default.nix:
```diff
@@ -0,0 +1,30 @@
+
+  nativeBuildInputs = [ libtool ];
+
+  patches = [ ./makefile.patch ];
```
Is this patch upstreamable?

veprbl replied:
Dated 2020-11-20, where does it come from? Please use `fetchpatch` where possible

OPNA2608 replied:
> Dated 2020-11-20, where does it come from?

https://github.com/archlinux/svntogit-community/blob/packages/pystring/trunk/makefile.patch

+1 on `fetchpatch`ing that instead.

rytone replied:
Replaced this patch with a `fetchpatch`ed pull request on the upstream repo that adds a CMake build.

OPNA2608 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -17511,7 +17513,9 @@ in
   openldap = callPackage ../development/libraries/openldap { };
 
-  opencolorio = callPackage ../development/libraries/opencolorio { };
+  opencolorio1 = callPackage ../development/libraries/opencolorio/1.x.nix { };
+  opencolorio2 = callPackage ../development/libraries/opencolorio { };
+  opencolorio = opencolorio1;
```
Would it make sense to make 2.x the default and pin everything that needs 1.x to `opencolorio1`?

SuperSandro2000 replied:
```suggestion
  opencolorio = callPackage ../development/libraries/opencolorio { };
  opencolorio_1 = callPackage ../development/libraries/opencolorio/1.x.nix { };
```

OPNA2608 commented on pkgs/development/libraries/opencolorio/default.nix:
```diff
@@ -1,47 +1,28 @@
 
-  outputs = [ "bin" "out" "dev" ];
+  nativeBuildInputs = [ cmake ];
+  buildInputs = [ expat libyamlcpp ilmbase pystring lcms2 python3Packages.pybind11 ];
```
> Right now, OCIO2 is compiled without GPU rendering support, but it should be possible to expose it as an optional feature if we want to.

Darwin build is broken for me because it detects GLUT in the regular system paths and attempts a build with GPU rendering support. Is there a CMake flag to explicitly disable GPU support?

If you want to work on enabling GPU support instead, the derivation needs `Carbon`, `GLUT` and `Cocoa` on Darwin to build.

```suggestion
  buildInputs = [
    expat libyamlcpp ilmbase pystring lcms2 python3Packages.pybind11
  ] ++ lib.optionals stdenv.hostPlatform.isDarwin [
    Carbon GLUT Cocoa
  ];
```

(In `all-packages.nix`, do `inherit (darwin.apple_sdk.frameworks) Carbon GLUT Cocoa;` for opencolorio2)

veprbl replied:
Can confirm this fixes the build
```diff
diff --git a/pkgs/development/libraries/opencolorio/default.nix b/pkgs/development/libraries/opencolorio/default.nix
index d9de24b7fa0..573d5f62cd6 100644
--- a/pkgs/development/libraries/opencolorio/default.nix
+++ b/pkgs/development/libraries/opencolorio/default.nix
@@ -1,4 +1,4 @@
-{ stdenv, lib, fetchFromGitHub, cmake, expat, libyamlcpp, ilmbase, pystring, lcms2, python3Packages }:
+{ stdenv, lib, fetchFromGitHub, cmake, expat, libyamlcpp, ilmbase, pystring, lcms2, python3Packages, Carbon, GLUT, Cocoa }:
 
 with lib;
 
@@ -14,7 +14,11 @@ stdenv.mkDerivation rec {
   };
 
   nativeBuildInputs = [ cmake ];
-  buildInputs = [ expat libyamlcpp ilmbase pystring lcms2 python3Packages.pybind11 ];
+  buildInputs = [
+    expat libyamlcpp ilmbase pystring lcms2 python3Packages.pybind11
+  ] ++ lib.optionals stdenv.hostPlatform.isDarwin [
+    Carbon GLUT Cocoa
+  ];
 
   cmakeFlags = [ "-DOCIO_INSTALL_EXT_PACKAGES=NONE" ];
 
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index 42306cf5b27..2877c8da84d 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -17514,7 +17514,9 @@ in
   openldap = callPackage ../development/libraries/openldap { };
 
   opencolorio1 = callPackage ../development/libraries/opencolorio/1.x.nix { };
-  opencolorio2 = callPackage ../development/libraries/opencolorio { };
+  opencolorio2 = callPackage ../development/libraries/opencolorio {
+    inherit (darwin.apple_sdk.frameworks) Carbon GLUT Cocoa;
+  };
   opencolorio = opencolorio1;
 
   opendmarc = callPackage ../development/libraries/opendmarc { };
```

rytone replied:
As there is no way to disable GPU rendering, I added the necessary dependencies for it on Linux and Darwin.

OPNA2608 reviewed (changes requested):
Welcome to Nixpkgs! I've left some comments with feedback/questions.

SuperSandro2000 commented on pkgs/development/libraries/opencolorio/1.x.nix:
```diff
@@ -0,0 +1,47 @@
+{ stdenv, lib, fetchFromGitHub, cmake, boost, pkg-config, lcms2, tinyxml, git }:
+
+with lib;
+
+stdenv.mkDerivation rec {
```
```suggestion
stdenv.mkDerivation rec {
```

SuperSandro2000 commented on pkgs/development/libraries/opencolorio/1.x.nix:
```diff
@@ -0,0 +1,47 @@
+    "-DUSE_EXTERNAL_TINYXML=ON"
+    # External libyamlcpp 0.6.* not compatible: https://github.com/imageworks/OpenColorIO/issues/517
+    "-DUSE_EXTERNAL_YAML=OFF"
+  ] ++ optional stdenv.isDarwin "-DOCIO_USE_BOOST_PTR=ON"
+    ++ optional (!stdenv.hostPlatform.isi686 && !stdenv.hostPlatform.isx86_64) "-DOCIO_USE_SSE=OFF";
```
```suggestion
  ] ++ lib.optional stdenv.isDarwin "-DOCIO_USE_BOOST_PTR=ON"
    ++ lib.optional (!stdenv.hostPlatform.isi686 && !stdenv.hostPlatform.isx86_64) "-DOCIO_USE_SSE=OFF";
```

SuperSandro2000 commented on pkgs/development/libraries/opencolorio/1.x.nix:
```diff
@@ -0,0 +1,47 @@
+  # TODO: Investigate whether git can be dropped: It's only used to apply patches
+  nativeBuildInputs = [ cmake pkg-config git ];
+
+  buildInputs = [ lcms2 tinyxml ] ++ optional stdenv.isDarwin boost;
```
```suggestion
  buildInputs = [ lcms2 tinyxml ] ++ lib.optional stdenv.isDarwin boost;
```

rastertail committed:
pystring: Use CMake build

Has a pull request open on the upstream repo, thus may be upstreamed in
the future as opposed to the previous makefile patch.

rastertail committed:
opencolorio2: Refactor dependencies

Python bindings and utility applications can now be disabled if desired.
GPU rendering is always enabled.

rastertail committed:
opencolorio: Replace with 2.x

All applications besides Blender have been pinned to 1.x

rastertail committed:
opencolorio_1: Remove top-level `with lib`

veprbl commented:
Result of `nixpkgs-review pr 127522` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>blender</li>
    <li>opencolorio</li>
    <li>opencolorio_1</li>
    <li>pystring</li>
  </ul>
</details>

InternetUnexplorer commented:
This isn't building for me :disappointed:. Maybe I'm doing something wrong? It builds fine in https://github.com/NixOS/nixpkgs/commit/9959bb87e828206c365d87fee72ff740a4a89dcc, but not in current master (https://github.com/NixOS/nixpkgs/commit/a7776f951e7d74743c141dcce8a1789dea3c76fe).

 `git bisect` says that it was broken in https://github.com/NixOS/nixpkgs/commit/d6234c2165eb670aaaa2071d976a66fcc3a082fb, but I don't know how helpful that is since that commit is a big merge.

Here is the build output of `nix build nixpkgs/a7776f951e7d74743c141dcce8a1789dea3c76fe#opencolorio --print-build-logs` on `x86_64-linux` (NixOS 21.11.20210625.e1f8852):

<details>
 <summary>Build output</summary>

```text
opencolorio> unpacking sources
opencolorio> unpacking source archive /nix/store/xns2qwzrd2ilsqrmd7818f22lgysbyx8-source
opencolorio> source root is source
opencolorio> patching sources
opencolorio> configuring
opencolorio> fixing cmake files...
opencolorio> cmake flags: -DCMAKE_FIND_USE_SYSTEM_PACKAGE_REGISTRY=OFF -DCMAKE_FIND_USE_PACKAGE_REGISTRY=OFF -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=ON -DCMAKE_BUILD_TYPE=Release -DCMAKE_SKIP_BUILD_RPATH=ON -DBUILD_TESTING=OFF -DCMAKE_INSTALL_LOCALEDIR=/nix/store/kvij6vg4iiz26j206xll43yhz1lg401f-opencolorio-2.0.1/share/locale -DCMAKE_INSTALL_LIBEXECDIR=/nix/store/kvij6vg4iiz26j206xll43yhz1lg401f-opencolorio-2.0.1/libexec -DCMAKE_INSTALL_LIBDIR=/nix/store/kvij6vg4iiz26j206xll43yhz1lg401f-opencolorio-2.0.1/lib -DCMAKE_INSTALL_DOCDIR=/nix/store/kvij6vg4iiz26j206xll43yhz1lg401f-opencolorio-2.0.1/share/doc/OpenColorIO -DCMAKE_INSTALL_INFODIR=/nix/store/kvij6vg4iiz26j206xll43yhz1lg401f-opencolorio-2.0.1/share/info -DCMAKE_INSTALL_MANDIR=/nix/store/kvij6vg4iiz26j206xll43yhz1lg401f-opencolorio-2.0.1/share/man -DCMAKE_INSTALL_OLDINCLUDEDIR=/nix/store/kvij6vg4iiz26j206xll43yhz1lg401f-opencolorio-2.0.1/include -DCMAKE_INSTALL_INCLUDEDIR=/nix/store/kvij6vg4iiz26j206xll43yhz1lg401f-opencolorio-2.0.1/include -DCMAKE_INSTALL_SBINDIR=/nix/store/kvij6vg4iiz26j206xll43yhz1lg401f-opencolorio-2.0.1/sbin -DCMAKE_INSTALL_BINDIR=/nix/store/kvij6vg4iiz26j206xll43yhz1lg401f-opencolorio-2.0.1/bin -DCMAKE_INSTALL_NAME_DIR=/nix/store/kvij6vg4iiz26j206xll43yhz1lg401f-opencolorio-2.0.1/lib -DCMAKE_POLICY_DEFAULT_CMP0025=NEW -DCMAKE_OSX_SYSROOT= -DCMAKE_FIND_FRAMEWORK=LAST -DCMAKE_STRIP=/nix/store/aigvlxh2aqjqd1wcnk1dr8cjml6dq846-binutils-2.35.1/bin/strip -DCMAKE_RANLIB=/nix/store/aigvlxh2aqjqd1wcnk1dr8cjml6dq846-binutils-2.35.1/bin/ranlib -DCMAKE_AR=/nix/store/aigvlxh2aqjqd1wcnk1dr8cjml6dq846-binutils-2.35.1/bin/ar -DCMAKE_C_COMPILER=gcc -DCMAKE_CXX_COMPILER=g++ -DCMAKE_INSTALL_PREFIX=/nix/store/kvij6vg4iiz26j206xll43yhz1lg401f-opencolorio-2.0.1 -DOCIO_INSTALL_EXT_PACKAGES=NONE
opencolorio> -- The CXX compiler identification is GNU 10.3.0
opencolorio> -- The C compiler identification is GNU 10.3.0
opencolorio> -- Detecting CXX compiler ABI info
opencolorio> -- Detecting CXX compiler ABI info - done
opencolorio> -- Check for working CXX compiler: /nix/store/x0xd1vp2w7n38g0hzflzy8p6hyl0dwic-gcc-wrapper-10.3.0/bin/g++ - skipped
opencolorio> -- Detecting CXX compile features
opencolorio> -- Detecting CXX compile features - done
opencolorio> -- Detecting C compiler ABI info
opencolorio> -- Detecting C compiler ABI info - done
opencolorio> -- Check for working C compiler: /nix/store/x0xd1vp2w7n38g0hzflzy8p6hyl0dwic-gcc-wrapper-10.3.0/bin/gcc - skipped
opencolorio> -- Detecting C compile features
opencolorio> -- Detecting C compile features - done
opencolorio> -- Setting C++ version to '11' as none was specified.
opencolorio> -- Performing Test COMPILER_SUPPORTS_CXX11
opencolorio> -- Performing Test COMPILER_SUPPORTS_CXX11 - Success
opencolorio> -- Found OpenGL: /nix/store/alk4sk1zagxpi0k7czkgs6fm63kmy13c-libGL-1.3.3/lib/libOpenGL.so  found components: OpenGL
opencolorio> -- Found GLEW: /nix/store/fqfs5iffsb6rdbamjfm5v02zkcph6wbv-glew-2.2.0-dev/include (found version "2.2.0")
opencolorio> -- Found GLUT: /nix/store/91b165nnnz5f3wnxg6dzly4fqahl4bw9-freeglut-3.2.1/lib/libglut.so
opencolorio> -- GLVND supported
opencolorio> -- Performing Test HAVE_SSE2
opencolorio> -- Performing Test HAVE_SSE2 - Success
opencolorio> -- Setting SOVERSION to '2.0' as none was specified.
opencolorio> CMake Error at /nix/store/hmnywaai8pbahlxriayriwfdviv8nzg4-expat-2.4.1-dev/lib/cmake/expat-2.4.1/expat.cmake:74 (message):
opencolorio>   The imported target "expat::expat" references the file
opencolorio>      "/nix/store/hmnywaai8pbahlxriayriwfdviv8nzg4-expat-2.4.1-dev/lib/libexpat.so.1.8.1"
opencolorio>   but this file does not exist.  Possible reasons include:
opencolorio>   * The file was deleted, renamed, or moved to another location.
opencolorio>   * An install or uninstall procedure did not complete successfully.
opencolorio>   * The installation package was faulty and contained
opencolorio>      "/nix/store/hmnywaai8pbahlxriayriwfdviv8nzg4-expat-2.4.1-dev/lib/cmake/expat-2.4.1/expat.cmake"
opencolorio>   but not all the files it references.
opencolorio> Call Stack (most recent call first):
opencolorio>   /nix/store/hmnywaai8pbahlxriayriwfdviv8nzg4-expat-2.4.1-dev/lib/cmake/expat-2.4.1/expat-config.cmake:35 (include)
opencolorio>   share/cmake/modules/Findexpat.cmake:33 (find_package)
opencolorio>   share/cmake/modules/FindExtPackages.cmake:29 (find_package)
opencolorio>   CMakeLists.txt:239 (include)
opencolorio> -- Configuring incomplete, errors occurred!
opencolorio> See also "/build/source/build/CMakeFiles/CMakeOutput.log".
error: builder for '/nix/store/w57bv43wn3a11v6bphsn2qbm5fian74s-opencolorio-2.0.1.drv' failed with exit code 1;
...
```
</details>

devhell commented:
I can confirm that it doesn't build, and it breaks Blender for me. @rytone could you please have a look?

tobiasBora commented:
Same for me, blender is broken now as I reported here https://github.com/NixOS/nixpkgs/issues/129092. The error I get is:
```
-- Setting SOVERSION to '2.0' as none was specified.
CMake Error at /nix/store/hmnywaai8pbahlxriayriwfdviv8nzg4-expat-2.4.1-dev/lib/cmake/expat-2.4.1/expat.cmake:74 (message):
  The imported target "expat::expat" references the file

     "/nix/store/hmnywaai8pbahlxriayriwfdviv8nzg4-expat-2.4.1-dev/lib/libexpat.so.1.8.1"

  but this file does not exist.  Possible reasons include:

  * The file was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and contained

     "/nix/store/hmnywaai8pbahlxriayriwfdviv8nzg4-expat-2.4.1-dev/lib/cmake/expat-2.4.1/expat.cmake"

  but not all the files it references.

Call Stack (most recent call first):
  /nix/store/hmnywaai8pbahlxriayriwfdviv8nzg4-expat-2.4.1-dev/lib/cmake/expat-2.4.1/expat-config.cmake:35 (include)
  share/cmake/modules/Findexpat.cmake:33 (find_package)
  share/cmake/modules/FindExtPackages.cmake:29 (find_package)
  CMakeLists.txt:239 (include)
```
